### PR TITLE
refactor!: abstract AwsSigner to support multiple backing implementations

### PR DESCRIPTION
**Pull request by ianbotsf**
## Issue \#

Closes [smithy-kotlin#616](https://github.com/awslabs/smithy-kotlin/issues/616)

## Description of changes

See main description in [companion PR](https://github.com/awslabs/smithy-kotlin/pull/626).

There's also a new `AwsSignerIntegration` which adds support for signing instead of doing it inline in `AwsHttpBindingProtocolGenerator`, necessary for adding a new service-level config parameter for signer implementation. The vast majority of the remaining changes are import/dependency changes.

Companion PR: [smithy-kotlin#626](https://github.com/awslabs/smithy-kotlin/pull/626)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-runtime/aws-endpoint/common/src/aws/sdk/kotlin/runtime/endpoint/AwsEndpointResolver.kt [hunk 1]
```diff
@@ -18,3 +22,10 @@ public fun interface AwsEndpointResolver {
      */
     public suspend fun resolve(service: String, region: String): AwsEndpoint
 }
+
+public fun AwsEndpointResolver.asSigningEndpointProvider(): SigningEndpointProvider = {
```
Comment (aajtodd) [inline aws-runtime/aws-endpoint/common/src/aws/sdk/kotlin/runtime/endpoint/AwsEndpointResolver.kt:26]: `InternalSdkApi`?

File: aws-runtime/aws-types/common/src/aws/sdk/kotlin/runtime/auth/credentials/CredentialsProvider.kt [hunk 2]
```diff
@@ -1,17 +0,0 @@
-/*
```
Comment (aajtodd) [inline aws-runtime/aws-types/common/src/aws/sdk/kotlin/runtime/auth/credentials/CredentialsProvider.kt:1]: it looks like only `RegionProvider` is left in `aws-types`. We should probably just relocate that and remove the `aws-types` project at this point.

Either aws core or aws-config seem reasonable to me.

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/AwsSignerIntegration.kt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import software.amazon.smithy.kotlin.codegen.core.CodegenContext
+import software.amazon.smithy.kotlin.codegen.core.RuntimeTypes
+import software.amazon.smithy.kotlin.codegen.integration.KotlinIntegration
+import software.amazon.smithy.kotlin.codegen.rendering.ClientConfigProperty
+import software.amazon.smithy.kotlin.codegen.rendering.ClientConfigPropertyType
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.ProtocolGenerator
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.ProtocolMiddleware
+import software.amazon.smithy.model.Model
+
+class AwsSignerIntegration : KotlinIntegration {
```
Comment (aajtodd) [inline codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/AwsSignerIntegration.kt:14]: technically we could move this to `smithy-kotlin` codegen at this point. we don't have to do it in this PR but maybe something to consider

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/customization/s3/S3SigningConfig.kt [hunk 4]
```diff
@@ -36,17 +36,17 @@ class S3SigningConfig : KotlinIntegration {
...
 }
 
 private class S3SigningMiddleware(signingServiceName: String) : AwsSignatureVersion4(signingServiceName) {
     override fun renderSigningConfig(op: OperationShape, writer: KotlinWriter) {
         super.renderSigningConfig(op, writer)
-        val sbht = AwsRuntimeTypes.Signing.AwsSignedBodyHeaderType
-        writer.addImport(sbht)
-        writer.write("signedBodyHeaderType = #T.X_AMZ_CONTENT_SHA256", sbht)
+        val sbh = RuntimeTypes.Auth.Signing.AwsSigningCommon.AwsSignedBodyHeader
+        writer.addImport(sbh)
```
Comment (aajtodd) [inline codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/customization/s3/S3SigningConfig.kt:48]: nit should be taken care of automagically now by `#T`

File: gradle.properties [hunk 5]
```diff
@@ -12,7 +12,7 @@ sdkVersion=0.14.4-SNAPSHOT
 smithyVersion=1.17.0
 smithyGradleVersion=0.5.3
 # smithy-kotlin codegen and runtime are versioned together
-smithyKotlinVersion=0.8.3
+smithyKotlinVersion=0.8.4-SNAPSHOT
```
Comment (aajtodd) [inline gradle.properties:15]: 0.8.6-SNAPSHOT now FYI

**Review by aajtodd — APPROVED** [review]

File: gradle.properties (hunk 5 above)
Comment (ianbotsf) [reply gradle.properties:15]: Yep yep, I'll have to do another merge-from-**main** today before I merge-to-**main**.

File: aws-runtime/aws-types/common/src/aws/sdk/kotlin/runtime/auth/credentials/CredentialsProvider.kt (hunk 2 above)
Comment (ianbotsf) [reply aws-runtime/aws-types/common/src/aws/sdk/kotlin/runtime/auth/credentials/CredentialsProvider.kt:1]: Everything that uses `RegionProvider` is in **aws-config** so moving it there.

**Pull request merged**